packages/firestore/src/remote/watch_change.ts

```diff
@@ -350,35 +338,252 @@ export class WatchChangeAggregator {
     });
   }
 
+  /** Resets a target after an existence filter mismatch. */
+  handleExistenceFilterMismatch(targetId: TargetId): void {
+    this.resetTarget(targetId);
+  }
+
   /**
-   * Record that we get a watch target add/remove by decrementing the number of
-   * pending target responses that we have.
+   * Converts the currently accumulated state into a remote event at the
+   * provided snapshot version. Resets the accumulated changes before returning.
    */
-  private recordTargetResponse(targetId: TargetId): void {
-    const newCount = (this.pendingTargetResponses[targetId] || 0) - 1;
-    if (newCount === 0) {
-      delete this.pendingTargetResponses[targetId];
+  createRemoteEvent(snapshotVersion: SnapshotVersion): RemoteEvent {
+    const targetChanges: { [targetId: number]: TargetChange } = {};
+
+    objUtils.forEachNumber(this.targetStates, (targetId, targetState) => {
+      const queryData = this.queryDataForActiveTarget(targetId);
+      if (queryData) {
+        if (targetState.shouldRaise) {
+          targetChanges[targetId] = targetState.toTargetChange();
+          targetState.clearChanges();
+        }
+
+        if (targetState.current && queryData.query.isDocumentQuery()) {
+          // Document queries for document that don't exist can produce an empty
+          // result set. To update our local cache, we synthesize a document
+          // delete if we have not previously received the document. This
+          // resolves the limbo state of the document, removing it from
+          // limboDocumentRefs.
+          //
+          // TODO(dimond): Ideally we would have an explicit lookup query
+          // instead resulting in an explicit delete message and we could
+          // remove this special logic.
+          const key = new DocumentKey(queryData.query.path);
+          if (
+            this.documentUpdates.get(key) === null &&
+            !this.hasSyncedDocument(targetId, key)
+          ) {
+            this.documentUpdates = this.documentUpdates.insert(
+              key,
+              new NoDocument(key, snapshotVersion)
+            );
+
+            // While we don't add the document to a target, we potentially
+            // need to mark it as a resolved limbo key. This requires us
+            // to add the document to the list of document target mappings.
+            this.ensureDocumentTargetMapping(key);
```

Comment:
I updated the code.

FWIW, the last version of this code didn't trigger limbo resolution before since we didn't write down that the target matched. There was nothing written down in the cache if we got no results back (which meant that the next time around we would not raise events from cache).

The test scenario that b/79422585 asked for is already handled by "Synthesizes deletes for missing document'.